### DISCUSSION TITLE
Configuring Strongswan with tpm2.0

### DISCUSSION THREAD

**Discussion by Mhamadkadiry** [strongswan/strongswan#2450] [Q&A] [answered]
I'm new to strongswan and tpm and I have a uni task to configure strongswan with tpm2.0 I tried to follow this  [documentation](https://develop.trustedcomputinggroup.org/2018/04/11/easy-tpm-2-0-access-with-the-strongswan-vpn-solution/), so I installed strongswan 5.9.10, tpm2-tss version 4.1.3 and tpm2-tools version 5.7. I managed to generate 

> 0x81010001   ECDSA Endorsement Key
> 0x81010002   RSA Endorsement Key
> 0x81010003   ECDSA Attestation Key
> 0x81010004   RSA Attestation Key

but when I try to generate X.509 endpoint certificate for the ECC attestation key using this command:
```
pki --req --keyid 0x81010003 --san koala.strongswan.org
    --dn "C=CH, O=strongSwan, CN=koala.strongswan.org"
    --outform pem > ak_ecc_req.pem
```

I get an output 
```
no PKCS#11 module found having a keyid 81:01:00:03
TPM 2.0 via TSS2 available
AIK signature algorithm is ECDSA with SHA256 hash
Smartcard PIN:
```
the certificate get generated but I'm worried about `no PKCS#11 module found having a keyid 81:01:00:03` because when I try to start strongsawn I get an error 
```
charon-systemd[487]: building CRED_CERTIFICATE - X509 failed, tried 6 builders
swanctl[609]: unable to load certificate: cert-tpm  

```
(PS: the token_ak_ecc is loaded successfully) and this is the sawnctl.conf 
```
connections {
  home {
    local {
      auth = pubkey
      cert-tpm {
        handle = 0x01800003
      }
      id = koala.strongswan.org
    }
  }
}
secrets {
  token_ak_ecc {
    handle = 0x81010003
  }
}
```
did I miss to install  some additional packages? I really appreciate any help because I'm still beginner and trying my first implementation :blush:

**Comment by Mhamadkadiry** [comment]
I'm using this script to install dependencies and packages did I miss something ?

```
#!/bin/bash
# Ipsec_setup.sh - Setup TPM and related software on Raspberry Pi

LOGFILE="/var/log/ipsec_setup.log"

# Function to log messages
log_message() {
    echo "$(date): $1" >> "$LOGFILE"
}

log_message "Starting TPM software setup"

echo "Dependencies for all TPM2 related software" | tee -a "$LOGFILE"
sudo apt -y update >> "$LOGFILE" 2>&1
sudo apt -y install doxygen libglib2.0-dev libdbus-1-dev automake libtool pkg-config gcc \
    libssl-dev libcurl4-gnutls-dev autoconf-archive libcmocka0 libcmocka-dev net-tools \
    build-essential git g++ m4 libgcrypt20-dev uthash-dev pandoc libusb-1.0-0-dev libjson-c-dev libgmp3-dev strongswan-pki \
    >> "$LOGFILE" 2>&1

log_message "Cloning, compiling and installing the latest TPM2-TSS version"
git clone https://github.com/tpm2-software/tpm2-tss.git >> "$LOGFILE" 2>&1
cd tpm2-tss
git checkout 4.1.3 >> "$LOGFILE" 2>&1
./bootstrap >> "$LOGFILE" 2>&1
./configure --with-udevrulesdir=/etc/udev/rules.d --with-udevrulesprefix=70- >> "$LOGFILE" 2>&1
make -j4 >> "$LOGFILE" 2>&1
sudo make install >> "$LOGFILE" 2>&1
sudo useradd --system --user-group tss >> "$LOGFILE" 2>&1
sudo udevadm control --reload-rules && sudo udevadm trigger >> "$LOGFILE" 2>&1
sudo ldconfig >> "$LOGFILE" 2>&1
cd ..

log_message "Cloning, compiling and installing the latest TPM2-ABRMD version"
git clone https://github.com/tpm2-software/tpm2-abrmd.git >> "$LOGFILE" 2>&1
cd tpm2-abrmd
git checkout 3.0.0 >> "$LOGFILE" 2>&1
./bootstrap >> "$LOGFILE" 2>&1
./configure --with-dbuspolicydir=/etc/dbus-1/system.d --with-systemdsystemunitdir=/lib/systemd/system \
    --with-systemdpresetdir=/lib/systemd/system-preset --datarootdir=/usr/share >> "$LOGFILE" 2>&1
make -j4 >> "$LOGFILE" 2>&1
sudo make install >> "$LOGFILE" 2>&1
sudo ldconfig >> "$LOGFILE" 2>&1
sudo pkill -HUP dbus-daemon >> "$LOGFILE" 2>&1
sudo systemctl daemon-reload >> "$LOGFILE" 2>&1
sudo systemctl enable tpm2-abrmd.service >> "$LOGFILE" 2>&1
sudo systemctl start tpm2-abrmd.service >> "$LOGFILE" 2>&1

log_message "DBus test"
dbus-send --system --dest=org.freedesktop.DBus --type=method_call --print-reply /org/freedesktop/DBus org.freedesktop.DBus.ListNames >> "$LOGFILE" 2>&1
cd ..

log_message "Cloning, compiling and installing the latest TPM2-Tools version"
git clone https://github.com/tpm2-software/tpm2-tools.git >> "$LOGFILE" 2>&1
cd tpm2-tools
git checkout 5.7 >> "$LOGFILE" 2>&1
./bootstrap >> "$LOGFILE" 2>&1
./configure >> "$LOGFILE" 2>&1
make -j4 >> "$LOGFILE" 2>&1
sudo make install >> "$LOGFILE" 2>&1
cd ..

# Install strongswan
sudo wget https://download.strongswan.org/strongswan-5.9.10.tar.bz2 >> "\$LOGFILE" 2>&1
sudo tar xjf strongswan-5.9.10.tar.bz2 >> "\$LOGFILE" 2>&1
cd strongswan-5.9.10
sudo ./configure --prefix=/usr --sysconfdir=/etc --enable-tpm --enable-tss-tss2 >> "\$LOGFILE" 2>&1
sudo make >> "\$LOGFILE" 2>&1
sudo make install >> "\$LOGFILE" 2>&1
```

**Comment by tobiasbrunner** [comment]
> the certificate get generated but I'm worried about `no PKCS#11 module found having a keyid 81:01:00:03` because when I try to start strongsawn I get an error

That's unrelated. It's logged by the _pkcs11_ plugin. If you don't use PKCS#11, don't build/load that plugin. Note that you haven't explicitly built it according to your script, however, you might still have it installed from a previous build or perhaps it gets pulled in by the `strongswan-pki` package, which you shouldn't install as the installed libraries and plugins will cause conflicts (i.e. make sure to uninstall all strongSwan-related packages before using your own build). Just build pki yourself via `--enable-pki`.

Also, that `pki --req` command you posted only generates a certificate request. You have to issue a certificate from it and then store that in the NV RAM of the TPM under the handle `0x01800003`. You might want to follow [this more complete tutorial](https://docs.strongswan.org/docs/5.9/tpm/tpm2.html).

**Reply by Mhamadkadiry** [reply]
Thank you it's working well! 
I have another point that I want to ask 
`sudo pki --req --type priv --keyid 0x81010003 --dn "C=CH, O=strongSec GmbH, OU=AK RSA, CN=edu.strongsec" --outform pem > ak_rsa_req.pem`
asks for `smartcard PIN` and I just want to skip that because I'm trying to automate this command, I've tried using `expect` 
```
expect <<EOF
spawn sudo pki --req --type priv --keyid 0x81010003 --dn "C=CH, O=strongSec GmbH, OU=AK RSA, CN=edu.strongsec" --outform pem > ak_rsa_req.pem
expect "Smartcard PIN:"
send "\r"
expect eof
EOF
```
to automatically press enter, which worked well but didn't generate the file `ak_rsa_req.pem`, so I'm wondering if we can just skip the question about the `smartcard PIN`, is it possible?

**Reply by tobiasbrunner** [reply, marked as answer]
As far as I understand, `spawn` redirects `stdout` of the process (connects it to `expect`). So you probably have to get the output from the `expect` command, while suppressing any output it generates itself. Maybe something like this:

```
expect > ak_rsa_req.pem <<EOF
log_user 0
spawn -noecho sudo pki --req --type priv --keyid 0x81010003 --dn "C=CH, O=strongSec GmbH, OU=AK RSA, CN=edu.strongsec" --outform pem
expect "Smartcard PIN:" {
  send "\r"
  log_user 1
}
expect eof
EOF
```
